lib/puppet/provider/service/service.rb

```diff
@@ -21,8 +21,23 @@ def restartcmd
 
   # A simple wrapper so execution failures are a bit more informative.
   def texecute(type, command, fof = true, squelch = false, combine = true)
+    # Set the process priority to 0 (or normal in Windows) so that services
+    # which are started as children of puppet will start with normal priority,
+    # rather than the priority of the puppet process itself.
+    if Facter.value(:osfamily) == 'windows'
+      priority = Process::NORMAL_PRIORITY_CLASS
+    else
+      priority = 0
+    end
```

Comment:
Could use `Puppet::Settings::PrioritySetting::PRIORITY_MAP[:normal]` to get the platform-specific priority value.

Reply:
That code relies on Puppet::Util::Platform.windows? which relies on `!!File::ALT_SEPARATOR` how would that work for the rspec tests? Since even on windows the specs for other init providers are run, e.g. systemd & openrc.

Reply:
Historically we've tried to run all of the tests on every platform, as it wasn't always clear which code executed on the agent vs server. For example, if a Windows provider requires a win32-specific gem without using puppet's feature system, them it can cause compilation to fail for all nodes, not just Windows. As a result, you'll see rspec tests for the `init`, `gentoo`, etc service providers running on Windows, and they often stub:

```ruby
allow(Puppet::Util::Platform).to receive(:windows?).and_return(false)
```

But now we generally exclude provider tests on platforms where it doesn't make sense, like https://github.com/puppetlabs/puppet/blob/97a68ba72ff158faab1cad0c784f4bb5f1967278/spec/unit/provider/service/openbsd_spec.rb#L3-L4

To your point about `PRIORITY_MAP`, yeah... I think it's fine to branch based on `Puppet::Util::Platform.windows?` and keep it how you have it, eg

```ruby
priority = Puppet::Util::Platform.windows? ? Process::NORMAL_PRIORITY_CLASS : 0
```

Reply:
@joshcooper I tried the approach you suggested, but that caused spec failures for Appveyor on windows. For instance:

```
Failures:
  1) Puppet::Type::Service::Provider::Gentoo #start should use the supplied start command if specified
     Failure/Error: execute(command, opts)
       Service[sshd](provider=gentoo) received :execute with unexpected arguments
         expected: (["/bin/foo"], {:combine=>true, :failonfail=>true, :override_locale=>false, :priority=>0, :squelch=>false})
              got: (["/bin/foo"], {:combine=>true, :failonfail=>true, :override_locale=>false, :priority=>32, :squelch=>false})
       Diff:
       @@ -2,6 +2,6 @@
         {:combine=>true,
          :failonfail=>true,
          :override_locale=>false,
       -  :priority=>0,
       +  :priority=>32,
          :squelch=>false}]
     # ./lib/puppet/provider/service/service.rb:36:in `texecute'
     # ./lib/puppet/provider/service/init.rb:177:in `texecute'
     # ./lib/puppet/provider/service/service.rb:50:in `ucommand'
     # ./lib/puppet/provider/service/base.rb:93:in `start'
     # ./spec/unit/provider/service/gentoo_spec.rb:83:in `block (3 levels) in <top (required)>'
     # util/rspec_runner:44:in `run'
     # util/rspec_runner:59:in `<main>
```

This happens because `priority = Puppet::Util::Platform.windows? ? Process::NORMAL_PRIORITY_CLASS : 0` returns 32 when running on windows even though we really want 0, as this is the Gentoo spec. Thoughts?

Reply:
I'd skip *nix service providers (gentoo, src, openrc, init, upstart, systemd) when running on Windows by adding the following to their respective top-level describe:

```
describe ..., unless: Puppet::Util::Platform.windows? do
```

Looks like you also need to remove the Facter `osfamily` expectation from some of the tests.

Reply:
@joshcooper that sounds reasonable, but it also seems like a fairly significant change to the puppet testing methodology. If that change is desired I think it should be a separate pull request. Should you or I file a separate bug report with the rationale for the change?

Following the creation of the bug report I could then craft a separate pull request to address that bug specifically.

Reply:
I submitted https://github.com/puppetlabs/puppet/pull/7534 to cleanup the specs

Reply:
> I submitted #7534 to cleanup the specs

awesome thanks, I'll rebase this work ontop of that PR, once it is merged in

Reply:
@joshcooper given that #7534 has been closed, any thoughts on the direction of this PR?

Reply:
To [@branan's point](https://github.com/puppetlabs/puppet/pull/7534#issuecomment-494958005), I think the execution API should present a platform agnostic way to specify the priority level (eg `:normal`), and have the different posix vs windows execution implementations map that to the appropriate platform specific value (e.g. `0` on posix or `Process::NORMAL_PRIORITY_CLASS` on windows).

Reply:
@joshcooper I pushed the platform specific values into the execution implementations, does that look better?

Reply:
@joshcooper any chance you have time for another review?